### PR TITLE
Automatically convert paths dropped on WSL instances

### LINKED ISSUE #331: Drag & Drop from File Explorer produces incorrect file paths with WSL
Dragging & dropping a file from Windows File Explorer into a Bash window produces the path formatted for Windows instead of Ubuntu.

Dragging the Desktop folder into a Bash window should produce the text /mnt/c/Users/acesnik/Desktop. Instead, it produces C:\Users\acesnik\Desktop.

Your Windows build number
Edition: Windows 10 Pro
Version: 1607
OS Build: 14393.576

See also https://github.com/Microsoft/WSL/issues/1501

### PR DESCRIPTION
**Pull request by petrsnm**
Drag and drop does not work for WSL because paths are pasted as windows
paths having incorrect path separator and path root.  This PR adds code
to correct the path in TerminalControl before pasting to WSL terminals.

One problem with this approach is that it assumes the default WSL
automount root of "/mnt". It would be possible to add a setting like
"WslDragAndDropMountRoot"... but I decided it if someone wants to change
automount location it would be simple enough just to create the "/mnt"
symlink in WSL.

## Validation
Couldn't find an obvious place to add a test.  Manually tested
cut-n-paste from following paths:
- "c:\"
- "c:\subdir"
- "c:\subdir\subdir"
- "\\wsl.localhost\<distro>"
- \\wsl.localhost\<distro>\subdir"

Closes #331

### CODE REVIEW

**Review by DHowett — DISMISSED** [review]
Huh. This PR seems to make some unexpected file mode changes.

image

It may be necessary to re-add the files that changed from `644` to `755` using `git add --chmod -x`

I'm also blocking to do a more in-depth review; sorry!

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -2271,6 +2271,35 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                     }
 
                     std::wstring fullPath{ item.Path() };
+
+                    // Fix path for WSL
+                    if (_settings.ProfileSource() == L"Windows.Terminal.Wsl")
+                    {
+                        static constexpr std::wstring_view wslPathPrefix{L"//wsl.localhost/"};
```
[the remark this answers is not in the document]
Comment (LuanVSO) [reply src/cascadia/TerminalControl/TermControl.cpp]: > run the prefix check twice
or we could do something like 
```c++
static std::wstring_view wslPathPrefix = onWindows11() ? L"//wsl.localhost/":L"//wsl$/";
```
and run it once

**Comment by DHowett** [comment]
Thanks for working on this, @petrsnm! Sorry for the delay in my reviewing it.

I love the direction it's headed, and this is definitely a feature that people will be happy about.

I'm a little iffy on the design, however, for one reason: `TerminalControl` was originally architected to be a reusable component that another app developer could take a dependency on. Through that lens, giving it knowledge of "profiles" and "sources" and how to translate paths--when the app developer using it may not have even connected it to a local process and likely doesn't have profiles or sources--falls short of that need. TC already has a few layers of indirection to support this which are visible in [how we handle pastes](https://github.com/microsoft/terminal/blob/7db7ba1ac9c1ee65f7b389f1893f985f10e68c17/src/cascadia/TerminalControl/ControlInteractivity.cpp#L171-L181), among other places.

I accept that this design is not obvious if you're jumping into the code out of nowhere, and that we should for sure make sure our documentation mentions our guiding principles. It's also a bit of a pain for a contributor to have to wrap their head around. _Because of that_, I'm not going to block your pull request over it. If you're not comfortable making those changes, or if me asking is overstaying my "maintainer welcome", or if you feel like you've done enough: what is here is perfectly acceptable and something that we can build on later.

Let me know :smile:

**Comment by petrsnm** [comment]
@DHowett,  I saw that there was already pre-paste fixup code for quoting paths with spaces in `TerminalControl._DragDropHandler`, so I assumed this was the right place to do pre-paste fixups.  It wasn't obvious to me that `TerminalControl' was meant to be reusable.  Incidentally, would quoting paths with spaces belong in `ControlInteractivity` also?  

It looks like it would not be that difficult to move the fix to `ControlInteractivity` but I'm currently buried in projects that I have to complete before Thanksgiving, and I don't think I'll have time to make a new pull request until December.  

I'm currently running a dev copy of Terminal because this PR makes a huge productivity difference for me.   Would it be possible to merge this PR , and then maybe I can come back and move the fix into `ControlInteractivity` when someone creates an issue that they can't reuse `TerminalControl` because they don't have "profiles" or "sources"?

**Review by zadjii-msft — APPROVED** [review]
FWIW we're cool merging this as-is. From an "architectural purity" standpoint, we should maybe move some logic around, but that's not something we'd totally expect an external contributor to do. The code is good, and the feature works, I can come back around later and move it 😋 Thanks again for doing this!

**Review by lhecker — APPROVED** [review]

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
Notes moved from PR body:

https://user-images.githubusercontent.com/6278445/139003088-f9d3e50b-f42a-4a96-a9ed-76ac91337cb7.mp4

**Comment by DHowett** [comment]
Thanks so much for doing this! I'm writing up the overall architecture plan for another comment, but I wanted to make sure I said thanks foremost :smile:

**Comment by loic** [comment]
My Terminal has 2 Ubuntu profiles. (I installed the Ubuntu app from Windows Store).

This fix works well with the Ubuntu profile that is `Hidden" by default (profile with a Tux icon and a command line of `wsl.exe -d Ubuntu`) but not with the default Ubuntu profile (profile with the Ubuntu logo icon and a command line of `ubuntu.exe`).

I haven't researched the concrete difference between `wsl.exe -d Ubuntu` and `ubuntu.exe`, but it would be nice if it worked with both, especially with the default profile. Anyone got that to work or knows what could be at play?

image

**Comment by zadjii-msft** [comment]
> I haven't researched the concrete difference between `wsl.exe -d Ubuntu` and `ubuntu.exe`, but it would be nice if it worked with both, especially with the default profile. Anyone got that to work or knows what could be at play?

Basically, the `ubuntu.exe` profile is one that Canonical ships as a "fragment extension", while the `wsl.exe -d Ubuntu` one is one that comes with the Terminal itself. Terminal only knows how to translate paths for our own WSL profiles, and can't be sure for the Canonical ones.

**Pull request merged**